src/components/tl/mlx5/mcast/tl_mlx5_mcast.h

```diff
@@ -175,18 +196,27 @@ struct pp_packet {
     uint32_t        psn;
     int             length;
     uintptr_t       context;
-    uintptr_t       buf;
+    uint32_t        packet_counter;
+    int             qp_id;
+    uintptr_t       buf; // buffer address, initialized once
 };
 
 struct mcast_ctx {
-    struct ibv_qp     *qp;
-    struct ibv_ah     *ah;
-    struct ibv_send_wr swr;
-    struct ibv_sge     ssg;
+    struct ibv_qp      *qp;
+    struct ibv_ah      *ah;
+    struct ibv_send_wr  swr;
+    struct ibv_sge      ssg;
+
     // RC connection info for supporing one-sided based relibality
     struct ibv_qp     **rc_qp;
     uint16_t           *rc_lid;
     union ibv_gid      *rc_gid;
+
+    // multiple mcast group
```

Comment:
This change will happen once all the allgather designs are in. I would rather not touch this struct as it will disturb the bcast design. At least we should first provide support for multiple mcast groups before making such a change. Eventually, we will have a single array of UD QPs that can be used for Allgather and Bcast designs.

Reply:
I am not sure to understand your answer. At least packing ibv_qp, ibv_ah, ibv_send_wr, ibv_sge into a struct sounds relevant: it makes the data more structured, reduces the duplication of the 4 lines, and prevents adding suffix "_list" just to differentiate the duplicated lines.

Since I am not getting the full picture but understand you don't want to make this change I am not insisting. Do as you think is best.

Reply:
Please let me clarify. When all the codes are in, the ```struct mcast_ctx``` will become like below. This is because we can use the same qp list for the bcast as well and we will not need individual pointers. 

from:
```
struct mcast_ctx {
    struct ibv_qp      *qp;
    struct ibv_ah      *ah;
    struct ibv_send_wr  swr;
    struct ibv_sge      ssg;

    // RC connection info for supporing one-sided based relibality
    struct ibv_qp     **rc_qp;
    uint16_t           *rc_lid;
    union ibv_gid      *rc_gid;

    // multiple mcast group
    struct ibv_qp      **qp_list;
    struct ibv_ah      **ah_list;
    struct ibv_send_wr  *swr_list;
    struct ibv_sge      *ssg_list;
};
````
to
````
struct mcast_ctx {
    // RC connection info for supporing one-sided based relibality
    struct ibv_qp     **rc_qp;
    uint16_t           *rc_lid;
    union ibv_gid      *rc_gid;

    // multiple mcast group
    struct ibv_qp      **qp_list;
    struct ibv_ah      **ah_list;
    struct ibv_send_wr  *swr_list;
    struct ibv_sge      *ssg_list;
};
````